Packs/HarfangLabEDR/Integrations/Hurukai/Hurukai.py

```diff
@@ -0,0 +1,2284 @@
+
+    already_fetched_previous = []
+    already_fetched_current = []
+
+    last_fetch = None
+    if last_run:
+        last_fetch = last_run.get('last_fetch', None)
+        already_fetched_previous = last_run.get('already_fetched', [])
+
+    if last_fetch is None:
+        # if missing, use what provided via first_fetch_time
+        last_fetch = first_fetch_time
+    else:
+        # otherwise use the stored last fetch
+        last_fetch = int(last_fetch)
+
+    if alert_status == 'ACTIVE':
+        status = ['new', 'probable_false_positive', 'investigating']
+    elif alert_status == 'CLOSED':
+        status = ['closed', 'false_positive']
+    else:
+        status = None
+
+    args = {
+        'ordering': '+alert_time',
+        'level': severity,
+        'limit': MAX_NUMBER_OF_ALERTS_PER_CALL,
+        'offset': 0
+    }  # type: Dict[str,Any]
+
+    if status:
+        args['status'] = ','.join(status)
+
+    if alert_type:
+        args['alert_type'] = alert_type
+
+    latest_created_time_us = 0
+
+    if last_fetch:
+        latest_created_time_us = int(last_fetch)
+        cursor = datetime.fromtimestamp(latest_created_time_us
+                                        / 1000000).replace(tzinfo=timezone.utc).strftime('%Y-%m-%dT%H:%M:%SZ')
+        args['alert_time__gte'] = cursor
+
+    incidents = []
+    total_number_of_alerts = 0
+
+    while True:
+
+        results = client._http_request(
+            method='GET',
+            url_suffix='/api/data/alert/alert/Alert/',
+            params=args
+        )
+
+        if 'count' in results and 'results' in results:
+            for alert in results['results']:
+                incident_created_time_us = int(datetime.timestamp(
+                    dateutil.parser.isoparse(alert.get('alert_time', '0'))) * 1000000)
+
+                # to prevent duplicates, we are only adding incidents with creation_time > last fetched incident
+                if last_fetch:
+                    if incident_created_time_us <= latest_created_time_us:
+                        continue
+
+                tags = alert.get('tags', [])
+                tactic = []
+                technique_id = []
+
+                for tag in tags:
+                    if tag.startswith('attack'):
+                        content = tag[7:]
+                        if content in TACTICS:
+                            tactic.append(TACTICS[content])
+                        elif content[0] == 't':
+                            technique_id.append(content)
+
+                incident = {
+                    'name': alert.get('rule_name', None),
+                    'occurred': alert.get('alert_time', None),
+                    'severity': SEVERITIES.index(alert.get('level', '').capitalize()) + 1,
+                    'rawJSON': json.dumps(alert)
+                }
+
+                alert_id = alert.get('id', None)
+                if alert_id not in already_fetched_previous:
```

Comment:
If an alert was fetched last time and has changed, we don't want to fetch it again with the change?

Reply:
We do not take the alert modification time into account for retrieving alerts. So changes in the EDR are not propagated to the XSOAR so far. 
But the analyst must decide whether his primary interface will be the XSOAR or the HarfangLab EDR.